src/webgpu/shader/validation/types/struct.spec.ts

```diff
@@ -98,11 +99,148 @@ struct S2 {
     t.expectCompileResult(t.params.target === 'i32', wgsl);
   });
 
-g.test('trailing_comma_after_member')
-  .desc('Test trailing comma after the last member declaration is allowed')
+interface StructureCase {
+  code: string;
+  valid: boolean;
+  f16?: boolean;
+}
+
+const kStructureCases: Record<string, StructureCase> = {
+  bool: {
+    code: `struct S { x : bool }`,
+    valid: true,
+  },
+  u32: {
+    code: `struct S { x : u32 }`,
+    valid: true,
+  },
+  i32: {
+    code: `struct S { x : i32 }`,
+    valid: true,
+  },
+  f32: {
+    code: `struct S { x : f32 }`,
+    valid: true,
+  },
+  f16: {
+    code: `struct S { x : f16 }`,
+    valid: true,
+    f16: true,
+  },
+
+  vec2u: {
+    code: `struct S { x : vec2u }`,
+    valid: true,
+  },
+  vec3i: {
+    code: `struct S { x : vec3i }`,
+    valid: true,
+  },
+  vec4f: {
+    code: `struct S { x : vec4f }`,
+    valid: true,
+  },
+  vec4h: {
+    code: `struct S { x : vec4h }`,
+    valid: true,
+    f16: true,
+  },
+
+  mat2x2f: {
+    code: `struct S { x : mat2x2f }`,
+    valid: true,
+  },
+  mat3x4h: {
+    code: `struct S { x : mat3x4h }`,
+    valid: true,
+    f16: true,
+  },
+
+  atomic_u32: {
+    code: `struct S { x : atomic<u32> }`,
+    valid: true,
+  },
+  atomic_i32: {
+    code: `struct S { x : atomic<i32> }`,
+    valid: true,
+  },
+
+  array_u32_4: {
+    code: `struct S { x : array<u32, 4> }`,
+    valid: true,
+  },
+  array_u32: {
+    code: `struct S { x : array<u32> }`,
+    valid: true,
+  },
+  array_u32_not_last: {
+    code: `struct S { x : array<u32>, y : u32 }`,
+    valid: false,
+  },
+  array_u32_override: {
+    code: `override o : u32;
+    struct S { x : array<u32, o> }`,
+    valid: false,
+  },
+
+  structure: {
+    code: `struct S { x : u32 }
+    struct T { x : S }`,
+    valid: true,
+  },
+  structure_structure_rta: {
+    code: `struct S { x : array<u32> }
+    struct T { x : S }`,
+    valid: false,
+  },
+
+  pointer: {
+    code: `struct S { x : ptr<function, u32> }`,
+    valid: false,
+  },
+
+  texture: {
+    code: `struct S { x : texture_2d<f32> }`,
+    valid: false,
+  },
+  sampler: {
+    code: `struct S { x : sampler }`,
+    valid: false,
+  },
+  sampler_comparison: {
+    code: `struct S { x : sampler_comparison }`,
+    valid: false,
+  },
+
+  many_members: {
+    code: `struct S {
+      m1 : u32,
+      m2 : i32,
+      m3 : vec4f,
+      m4 : array<u32, 8>,
+      m5 : array<f32>
+    }`,
+    valid: true,
+  },
+
+  trailing_comma: {
+    code: `struct S { x : u32, }`,
+    valid: true,
+  },
+};
```

Comment:
Additional suggested failure cases:
* Empty struct
* Struct with the same name as another struct, alias, function.